src/System.IO.FileSystem/src/System/IO/UnixFileSystem.cs

```diff
@@ -344,104 +354,223 @@ public override IEnumerable<FileSystemInfo> EnumerateFileSystemInfos(string full
                             }
-                            if (searchOption == SearchOption.AllDirectories)
+                            bool matchesSearchPattern =
+                                (isFile || isDir) &&
+                                Interop.libc.fnmatch(_searchPattern, name, Interop.libc.FnmatchFlags.None) == 0;
+
+                            // Yield the result if the user has asked for it.  In the case of directories,
+                            // always explore it by pushing it onto the stack, regardless of whether
+                            // we're returning directories.
+                            if (isDir && !ShouldIgnoreDirectory(name))
                             {
-                                if (toExplore == null)
+                                if (_includeDirectories && matchesSearchPattern)
+                                {
+                                    yield return _translateResult(Path.Combine(dirPath.UserPath, name), /*isDirectory*/true);
+                                }
+                                if (_searchOption == SearchOption.AllDirectories)
                                 {
-                                    toExplore = new Stack<string>();
+                                    if (toExplore == null)
+                                    {
+                                        toExplore = new Stack<PathPair>();
+                                    }
+                                    toExplore.Push(new PathPair(Path.Combine(dirPath.UserPath, name), Path.Combine(dirPath.FullPath, name)));
                                 }
-                                toExplore.Push(fullNewName);
+                            }
+                            else if (isFile && _includeFiles && matchesSearchPattern)
+                            {
+                                yield return _translateResult(Path.Combine(dirPath.UserPath, name), /*isDirectory*/false);
                             }
                         }
-                        else if (isFile && includeFiles && matchesSearchPattern)
+                    }
+                    finally
+                    {
+                        // Close the directory enumerator
+                        if (dirHandle != null && !dirHandle.IsClosed) // TODO: Remove once exception handling bug is fixed
                         {
-                            yield return translateResult(fullNewName, /*isDirectory*/false);
+                            dirHandle.Dispose();
+                            dirHandle = null;
                         }
                     }
+
+                    if (toExplore != null && toExplore.Count > 0)
+                    {
+                        // Open the next directory.
+                        dirPath = toExplore.Pop();
+                        dirHandle = OpenDirectory(dirPath.FullPath);
+                    }
+                }
+            }
+
+            private struct PathPair
```

Comment:
Both values are strings, paths, and potentially identical or at least similar, so with Key/Value or Item1/Item2 names, it'd be very easy to accidentally use the wrong one.